File: 程序员面试宝典/2.5.链表求和.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        new_head = ListNode(0)
        p = new_head

        flag = 0

        while l1 and l2:
            total = l1.val + l2.val + flag
            flag = total // 10
            val = total % 10

            node = ListNode(val)
            p.next = node
            p = p.next

            l1 = l1.next
            l2 = l2.next

        while l1:
            total = l1.val + flag
            flag = total // 10
            val = total % 10

            node = ListNode(val)
            p.next = node
            p = p.next

            l1 = l1.next

        while l2:
            total = l2.val + flag
            flag = total // 10
            val = total % 10

            node = ListNode(val)
            p.next = node
            p = p.next

            l2 = l2.next

        if flag:
            node = ListNode(flag)
            p.next = node
            p = p.next

        return new_head.next

File: 程序员面试金典/2.5.链表求和.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x)
        : val(x)
        , next(NULL)
    {
    }
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2)
    {
        ListNode* p = l1;
        ListNode* q = l2;
        ListNode* new_head = new ListNode(0);
        ListNode* cur = new_head;
        int carry = 0;

        while (p || q || carry != 0) {
            int val = 0;

            if (p) {
                val += p->val;
                p = p->next;
            }
            if (q) {
                val += q->val;
                q = q->next;
            }
            if (carry != 0) {
                val += carry;
            }

            ListNode* node = new ListNode(val % 10);
            cur->next = node;
            cur = cur->next;
            carry = val / 10;
        }

        return new_head->next;
    }
};